src/plot/ChunkCompressor.hpp

```diff
@@ -0,0 +1,344 @@
+    static std::vector<uint8_t> compress(const std::vector<uint8_t>& deltas,
+                                         const std::vector<uint64_t>& stubs,
+                                         uint8_t stub_bits)
+    {
+        if (deltas.size() != stubs.size()) {
+            throw std::invalid_argument("ChunkCompressor::compress: deltas and stubs size mismatch");
+        }
+
+        const uint32_t num_values = static_cast<uint32_t>(deltas.size());
+        if (num_values == 0) {
+            // Encode an empty chunk with zero sizes.
+            std::vector<uint8_t> chunk;
+            chunk.reserve(12);
+            append_u32(chunk, 0); // num_values
+            append_u32(chunk, 0); // fse_size
+            append_u32(chunk, 0); // stub_bytes_size
+            return chunk;
+        }
+
+        if (stub_bits == 0 || stub_bits >= 64) {
+            throw std::invalid_argument("ChunkCompressor::compress: stub_bits must be in [1, 63]");
+        }
+
+        // 1) FSE-compress the deltas
+        size_t srcSize = deltas.size();
+        size_t maxDst  = FSE_compressBound(srcSize);
+        std::vector<uint8_t> fse_data(maxDst);
+
+        size_t cSize = FSE_compress(fse_data.data(), maxDst,
+                                    deltas.data(), srcSize);
+        if (FSE_isError(cSize)) {
+            throw std::runtime_error("ChunkCompressor::compress: FSE_compress failed");
+        }
+        fse_data.resize(cSize);
+        const uint32_t fse_size = static_cast<uint32_t>(cSize);
+
+        // 2) Bit-pack stubs into bytes
+        std::vector<uint8_t> stub_bytes = packStubs(stubs, stub_bits);
+        const uint32_t stub_bytes_size = static_cast<uint32_t>(stub_bytes.size());
+
+        // 3) Build chunk blob
+        std::vector<uint8_t> chunk;
+        chunk.reserve(12 + fse_size + stub_bytes_size);
+
+        append_u32(chunk, num_values);
+        append_u32(chunk, fse_size);
+        append_u32(chunk, stub_bytes_size);
+
+        // fse data
+        chunk.insert(chunk.end(), fse_data.begin(), fse_data.end());
+        // stub data
+        chunk.insert(chunk.end(), stub_bytes.begin(), stub_bytes.end());
+
+        return chunk;
+    }
+
+    // Decompress a single chunk
+    // - chunk: bytes produced by compress()
+    // - stub_bits: same as used during compression
+    // Output:
+    // - out_deltas: resized to num_values, filled with deltas
+    // - out_stubs:  resized to num_values, filled with stub values
+    static void decompress(const std::vector<uint8_t>& chunk,
+                           uint8_t stub_bits,
+                           std::vector<uint8_t>& out_deltas,
+                           std::vector<uint64_t>& out_stubs)
+    {
+        if (chunk.size() < 12) {
+            throw std::runtime_error("ChunkCompressor::decompress: chunk too small");
+        }
+        if (stub_bits == 0 || stub_bits >= 64) {
+            throw std::invalid_argument("ChunkCompressor::decompress: stub_bits must be in [1, 63]");
```

Comment:
can these failures happen because of untrusted input, or would these indicate a programmer error?
It kind of looks like the latter. If so, I think these should be `assert()`s. (same in `compress()`)